---
title: "SARS-CoV-2 Variants Coinfection"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: 
  html_document:
    theme: cosmo
    toc: yes
    toc_float:
      collapsed: true
---
## Lineage assignment
 
### Pangolin and Nextclade 
 Lineage assigment table using Pangolin and Nextclade of all samples.
 ```{r, echo=FALSE}
 kable(params$lineage_assignment, "simple")
 ```

### Freyja
  The Freyja plot shows the relative SARS-CoV-2 lineage abundance in each sample. Samples suspect for coinfection will illustrate mixed prevalence of two or more variants. 
 ![](`r params$freyja_plot`)

### VirStrain
 
 Sample/s suspect for coinfection were run against a database containing representative SARS-CoV-2 sequences. The VirStrain result shows the possible lineage assignment for the sample/s.

## Bammix flags for nucleotide mixture
  The figure below shows the Bammix plot for nucleotide mixtures of the flagged sample/s. Multiple sites containing heterogenous bases crossing the threshold of >20% mixture will be apparent. 
 ![](`r params$bammix_plot`)

## Alternative allele fraction

### For each lineage-defining mutations
 The alternative allele fraction (AAF) determines the number of reads supporting each variant detected in the sample/s. The AAF per mutation of sample/s flagged for SARS-CoV-2 coinfection are shown in the figure. Delta, Omicron, and shared mutations are represented by blue, pink, and purple, respectively. 
 ![](`r params$aafplot_mut`)

### For each amplicon
  The AAF for each amplicon is plotted by getting the average allele fraction for Delta and Omicron defining mutation within the range of that amplicon.
 ![](`r params$aafplot_amp`)
  
## References {.tabset .tabset-fade}
### Pipeline
 The tools used in this [pipeline](https://github.com/lanadelrea/CoPi) are as follows: 

 - [Pangolin](https://github.com/cov-lineages/pangolin) 

 - [Nextclade](https://github.com/nextstrain/nextclade)

 - [Freyja](https://github.com/andersen-lab/Freyja)

 - [VirStrain](https://github.com/liaoherui/VirStrain)

 - [Bammix](https://github.com/chrisruis/bammix)

 - Custom scripts adopted from [Bolze et al.](doi: 10.1016/j.medj.2022.10.002) 

### Session Info
 ```{r, echo=FALSE}
 sessionInfo()
 ```